### PR TITLE
Issue #318 - New apis for index query totals.  testRawQueries() unit test updated.

### PR DESCRIPTION
**Pull request by holleyism**
Signed-off-by: Adam Holley <holleya@us.ibm.com>

### CODE REVIEW

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 1]
```diff
@@ -509,42 +509,57 @@ private void commitDeletes(String collectionName, List<String> deleteIds) throws
...
             throw new PermanentBackendException(e);
         }
+    }
+    
+    @Override
+    public Iterable<RawQuery.Result<String>> query(RawQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx) throws BackendException {
+    	String collection = query.getStore();
+    	String keyIdField = getKeyFieldId(collection);
+    	List<RawQuery.Result<String>> result;
+    	QueryResponse response = runCommonQuery(query, informations, tx, collection, keyIdField);
```
Comment (amcp) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: final

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 2]
```diff
@@ -509,42 +509,57 @@ private void commitDeletes(String collectionName, List<String> deleteIds) throws
...
+        if (logger.isDebugEnabled())
+            logger.debug("Executed query [{}] in {} ms", query.getQuery(), response.getElapsedTime());
+
+        int totalHits = response.getResults().size();
+        if (!query.hasLimit() && totalHits >= maxResults) {
+            logger.warn("Query result set truncated to first [{}] elements for query: {}", maxResults, query);
+        }
+        result = new ArrayList<RawQuery.Result<String>>(totalHits);
+        for (SolrDocument hit : response.getResults()) {
```
Comment (amcp) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: response.getResults().stream().map(to a RawQuery.Result).collect(Collectors.toList())

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 3]
```diff
@@ -509,42 +509,57 @@ private void commitDeletes(String collectionName, List<String> deleteIds) throws
...
+            double score = Double.parseDouble(hit.getFieldValue("score").toString());
+            result.add(new RawQuery.Result<String>(hit.getFieldValue(keyIdField).toString(), score));
+        }
         return result;
     }
 
+    @Override
+    public Long totals(RawQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx) throws BackendException {
+        Long result;
+        String collection = query.getStore();
```
Comment (amcp) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: final

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 4]
```diff
@@ -509,42 +509,57 @@ private void commitDeletes(String collectionName, List<String> deleteIds) throws
...
+        String keyIdField = getKeyFieldId(collection);
```
Comment (amcp) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: final

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java [hunk 5]
```diff
@@ -919,13 +918,28 @@ else if (Geoshape.class.isAssignableFrom(datatype)) {
...
 
     @Override
+    public Long totals(RawQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx) throws BackendException {
+        final ElasticSearchResponse response = runCommonQuery(query, informations, tx);
+        log.debug("Executed query [{}] in {} ms", query.getQuery(), response.getTook());
+        return response.getTotal();
+    }
+
+    
+
```
Comment (sjudeng) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java]: Remove extra newlines

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java [hunk 6]
```diff
@@ -612,14 +612,15 @@ public IndexQuery getQuery(final MixedIndexType index, final Condition condition
 //        }
 //    }
 
-    public Iterable<RawQuery.Result> executeQuery(IndexQueryBuilder query, final ElementCategory resultType,
-                                                  final BackendTransaction backendTx, final StandardJanusGraphTx transaction) {
+    //
+    // Common code used by executeQuery and executeTotals
+    private MixedIndexType createMixedIndex(IndexQueryBuilder query, final ElementCategory resultType,
+            final BackendTransaction backendTx, final StandardJanusGraphTx transaction, StringBuffer qB) {
```
Comment (holleyism) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java]: I agree.  I've updated and cleaned it up a little.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java [hunk 7]
```diff
@@ -666,11 +668,17 @@ else if (transaction.containsRelationType(keyname)) {
...
         log.info("Converted query string with {} replacements: [{}] => [{}]",replacements,query.getQuery(),queryStr);
-        RawQuery rawQuery=new RawQuery(index.getStoreName(),queryStr,query.getParameters());
+        return queryStr;
+    }
+    
+    public Iterable<RawQuery.Result> executeQuery(IndexQueryBuilder query, final ElementCategory resultType,
+                                                  final BackendTransaction backendTx, final StandardJanusGraphTx transaction) {
+        final MixedIndexType index = getMixedIndex(query.getIndex(), transaction);
+        String queryStr = createQueryString(query, resultType, transaction, index);
+        RawQuery rawQuery = new RawQuery(index.getStoreName(),queryStr,query.getParameters());
```
Comment (sjudeng) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java]: final queryStr and rawQuery

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java [hunk 8]
```diff
@@ -682,6 +690,15 @@ else if (transaction.containsRelationType(keyname)) {
         });
     }
 
+    public Long executeTotals(IndexQueryBuilder query, final ElementCategory resultType,
+            final BackendTransaction backendTx, final StandardJanusGraphTx transaction) {
+        final MixedIndexType index = getMixedIndex(query.getIndex(), transaction);
+        String queryStr = createQueryString(query, resultType, transaction, index);
+        RawQuery rawQuery = new RawQuery(index.getStoreName(),queryStr,query.getParameters());
```
Comment (sjudeng) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java]: final queryStr and rawQuery

**Comment by amcp** [comment]
Also please squash

**Comment by pluradj** [comment]
Fixes #318

**Pull request merged**